src/net/rpc.rs

```diff
@@ -1532,23 +1531,75 @@ impl ConversationHttp {
         http: &mut StacksHttp,
         fd: &mut W,
         req: &HttpRequestType,
-        tip_opt: Option<&StacksBlockId>,
+        tip_req: &TipRequest,
         sortdb: &SortitionDB,
-        chainstate: &StacksChainState,
+        chainstate: &mut StacksChainState,
     ) -> Result<Option<StacksBlockId>, net_error> {
-        match tip_opt {
-            Some(tip) => Ok(Some(*tip).clone()),
-            None => match chainstate.get_stacks_chain_tip(sortdb)? {
+        match tip_req {
+            TipRequest::UseLatestUnconfirmedTip => {
+                let unconfirmed_chain_tip_opt = match &mut chainstate.unconfirmed_state {
+                    Some(unconfirmed_state) => {
+                        if unconfirmed_state.is_readable() {
+                            // Check if underlying MARF trie exists before returning unconfirmed chain tip
+                            let trie_exists = match unconfirmed_state
+                                .clarity_inst
+                                .trie_exists_for_block(&unconfirmed_state.unconfirmed_chain_tip)
+                            {
+                                Ok(res) => res,
+                                Err(e) => {
+                                    let response_metadata = HttpResponseMetadata::from(req);
+                                    warn!("Failed to load Stacks chain tip; error checking underlying trie");
+                                    let response = HttpResponseType::ServerError(
+                                        response_metadata,
+                                        format!("Failed to load Stacks chain tip: {:?}", e),
+                                    );
+                                    return response.send(http, fd).and_then(|_| Ok(None));
+                                }
+                            };
+
+                            if trie_exists {
+                                Some(unconfirmed_state.unconfirmed_chain_tip)
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        }
+                    }
+                    None => None,
+                };
+
+                if let Some(unconfirmed_chain_tip) = unconfirmed_chain_tip_opt {
+                    Ok(Some(unconfirmed_chain_tip))
+                } else {
+                    match chainstate.get_stacks_chain_tip(sortdb)? {
+                        Some(tip) => Ok(Some(StacksBlockHeader::make_index_block_hash(
+                            &tip.consensus_hash,
+                            &tip.anchored_block_hash,
+                        ))),
+                        None => {
+                            let response_metadata = HttpResponseMetadata::from(req);
+                            warn!("Failed to load Stacks chain tip");
+                            let response = HttpResponseType::ServerError(
```

Comment:
This would also be a 404, not a 500.  A stacks node that hasn't downloaded any blocks yet is not in an error state.